[webkitpy] Avoid autoinstalling libraries in init

JonWBedard opened the pull request:
#### 5f9e420cedb1ad9e41602f43e9423e10c9a1b71a
<pre>
[webkitpy] Avoid autoinstalling libraries in init
<a href="https://bugs.webkit.org/show_bug.cgi?id=258366">https://bugs.webkit.org/show_bug.cgi?id=258366</a>
rdar://111120365

Reviewed by Elliott Williams.

Re-organize webkitscmpy and webkitbugspy to avoid invoking the autoinstaller
in webkitpy&apos;s __init__.py.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
Lazy-evaluate requests import.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/user.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/call_by_need.py:
(CallByNeed.__getattribute__): Attempt to use underlying object&apos;s attributes
when no type is defined.
(CallByNeed.__call__): If the value is callable, call it.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit_classifier.py:
(CommitClassifier.LineFilter.fuzzy): Move rapidfuzz import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/scm.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
Move mock import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/svn.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/svn.py:
Move xmldict import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/install_git_lfs.py:
Lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.Filters.fuzzy): Move rapidfuzz import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
Lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/pull_request.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
Use webkitcorepy instead of six, lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/scm.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/svn.py:
Lazy import requests and xmltodict.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/scm_base.py:
Use webkitcorepy instead of six.

Canonical link: <a href="https://commits.webkit.org/265377@main">https://commits.webkit.org/265377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fc6dc6fc1d368e84ca597941a3bf6cee03249f0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/10759 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10979 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/11279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/12408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/10317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10774 "Failed to checkout and rebase branch from PR 15157") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13353 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/10955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/13224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/10919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/11827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/11279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12811 "Failed to checkout and rebase branch from PR 15157") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/10863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/9121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/11279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/10192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/11279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13106 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/10330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/10664 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/9488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/11279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/13761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1205 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/10191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple reviewed (commented):
Makes sense, though if we can require Python 3.5+ (which is a big if, since I think this code still has to work on 2.7…) we could use [`importlib.util.LazyLoader`]( https://docs.python.org/3/library/importlib.html#importlib.util.LazyLoader) and replace these import statements with something like:

```python
from webkitcorepy import lazy_import 
requests = lazy_import("requests")
```

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
```diff
@@ -735,6 +734,8 @@ def __init__(
         )
 
     def __enter__(self):
+        from mock import patch
+
```
Something to keep in mind for the near future: We can almost certainly switch to using `unittest.mock` once we require Python 3.3+

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
Out of curiosity, we _don't_ import mocks from production code, right?

JonWBedard replied:
No, but we do export the `mock` structures, so we trigger the top level imports.

Many consumers of `webkitscmpy` need `webkitscmpy`'s mocks in their own unit tests. `reporelay` and `resultsdbpy` both do this.

emw-apple commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/call_by_need.py:
```diff
@@ -31,11 +31,7 @@ def __getattribute__(self, name):
-                "'{}' object has no attribute '{},' and doesn't define a target type".format(type(self).__name__, name),
-            )
-        if name in dir(typ):
+        if typ is None or name in dir(typ):
```
IIUC, this change removes the check that the object returned by `self.value` actually has an attribute named `name`? That seems fine.

JonWBedard replied:
This change removes the check that the declared type of the object `CallByNeed` is supposed to return has an attribute named `name`, if no `type` was specified.

My assumption here is that if a caller isn't careful enough to specify the expected return type, the caller is willing to accept the consequence of an attribute error coming from Python.

JonWBedard committed:
[webkitpy] Avoid autoinstalling libraries in init
https://bugs.webkit.org/show_bug.cgi?id=258366
rdar://111120365

Reviewed by Elliott Williams.

Re-organize webkitscmpy and webkitbugspy to avoid invoking the autoinstaller
in webkitpy's __init__.py.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
Lazy-evaluate requests import.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/user.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/call_by_need.py:
(CallByNeed.__getattribute__): Attempt to use underlying object's attributes
when no type is defined.
(CallByNeed.__call__): If the value is callable, call it.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit_classifier.py:
(CommitClassifier.LineFilter.fuzzy): Move rapidfuzz import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/scm.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
Move mock import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/svn.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/svn.py:
Move xmldict import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/install_git_lfs.py:
Lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pickable.py:
(Pickable.Filters.fuzzy): Move rapidfuzz import into function call.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
Lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/pull_request.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
Use webkitcorepy instead of six, lazy import requests.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/scm.py:
Use webkitcorepy instead of six.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/svn.py:
Lazy import requests and xmltodict.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/scm_base.py:
Use webkitcorepy instead of six.

Canonical link: https://commits.webkit.org/265377@main
